sql: introduce CREATE TENANT syntax

adityamaru opened the pull request:
This change introduces:

CREATE TENANT <tenant-name>

The query  automatically assigns the tenant the first
available ID. This could be an ID that was previously
used by a tenant that has been dropped and GC'ed.

Fixes: https://github.com/cockroachdb/cockroach/issues/91235

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/91595)
<!-- Reviewable:end -->

adityamaru committed:
logictest: rename tenant to tenant_builtins

Release note: None

adityamaru committed:
sql: introduce CREATE TENANT $1 grammar

This change introduces:

CREATE TENANT <tenant-name>

The query  automatically assigns the tenant the first
available ID. This could be an ID that was previously
used by a tenant that has been dropped and GC'ed.

Fixes: #91235

Release note: None

adityamaru commented:
actually hold off on the review for this, i want to try something

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91595)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @knz, and @lidorcarmel)

___
*[`-- commits` line 34 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91595#-NGV3q422oYIuG-1w63Q):*
i just saw this PR fly by, so i might be missing some context, but i had a question. adding `CREATE TENANT` to the release notes seems like it would also require a lot of extra background docs to explain what tenants are. so far the docs team has explicitly avoided documenting multitenancy, so i just want to make sure that documentation work already tracked somewhere for that team


<!-- Sent from Reviewable.io -->

adityamaru commented:
>  it would also require a lot of extra background docs

You're right! I discussed this with the team and we'll default to no release notes for now. We might add some private preview docs closer to the release data but that can be done independently. Thanks for flagging.

adityamaru commented:
@lidorcarmel @knz this is ready for a look. I added a `type TenantName string` to strongly type the string params we're passing around. Let me know what you think!

ajstorm commented:
>You're right! I discussed this with the team and we'll default to no release notes for now. 

Does it makes sense to also skip the documentation on the `CREATE TENANT` statement until we have confirmation that we want this exposed?

rafiss commented:
Yeah we need to make sure this doesn't show up in the auto-generated docs + diagrams

lidorcarmel commented on docs/generated/redact_safe.md:
```diff
@@ -29,6 +29,7 @@ pkg/roachpb/metadata.go | `ReplicaType`
 pkg/roachpb/metadata.go | `StoreID`
 pkg/roachpb/method.go | `Method`
 pkg/roachpb/tenant.go | `TenantID`
+pkg/roachpb/tenant.go | `TenantName`
```
are we sure this is safe for reporting? looking briefly in this file, I think this would be the only string here that is generated by the user, which is why I ask..

knz replied:
indeed, that doesn't seem right

adityamaru replied:
good catch, it was a copy pasta of TenantID that is marked `RedactSafe`. Removing.

lidorcarmel commented on pkg/sql/create_tenant.go:
```diff
@@ -0,0 +1,37 @@
+	name roachpb.TenantName
+}
+
+func (p *planner) CreateTenantNode(_ context.Context, n *tree.CreateTenant) (planNode, error) {
```
wondering, is this where we should verify things like name uniqueness and maybe permissions? or should we rely on startExec to fail?

knz replied:
- name uniqueness is ensured via the UNIQUE index on the column.
- permission is ensured inside the `CreateTenant` function.

stevendanna reviewed (approved):
Nice to see this making progress.

Looks like others have caught the important things. I've just left a few nitpicks.

adityamaru commented on pkg/sql/logictest/testdata/logic_test/tenant:
```diff
@@ -5,201 +5,34 @@ SELECT id, active, length(info), name FROM system.tenants ORDER BY id
 user testuser
 
 statement error only users with the admin role are allowed to create tenant
-SELECT crdb_internal.create_tenant(314)
```
done

adityamaru commented on pkg/ccl/multitenantccl/tenantcostclient/tenant_side_test.go:
```diff
@@ -922,7 +922,7 @@ func TestSQLLivenessExemption(t *testing.T) {
 
 	// Create a tenant with ridiculously low resource limits.
 	host := sqlutils.MakeSQLRunner(hostDB)
-	host.Exec(t, "SELECT crdb_internal.create_tenant($1)", tenantID.ToUint64())
```
I _think_ this will need a little more work because we only have certs for tenant 10, 11, and 12 and a `CREATE TENANT` will create a tenant at ID 2.

stevendanna replied:
Gotcha, thanks!

adityamaru commented:
> Yeah we need to make sure this doesn't show up in the auto-generated docs + diagrams

So I see that `ALTER TENANT` has all the generated diagrams but isn't on our public facing docs. I'm working with @kathancox to ensure this follows the same pattern and doesn't show up anywhere.

adityamaru commented:
TFTR!

bors r=stevendanna,lidorcarmel
